### PR TITLE
obs-studio: use wrapobs and the new plugin layout

### PR DESCRIPTION
**Pull request by 06kellyjac**
### Description

The packaging in nixpkgs for obs plugins has changed and there's a `wrapOBS` function.
The name of the plugins has also changed so the example needed updating to reflect that.

Related: https://github.com/NixOS/nixpkgs/pull/125308

This PR is done in the style of the `neovim` module that has `package` and a hidden `finalPackage`

I have not added a news update as this is more of a behind-the-scenes change but I can if you'd like.

Is there any issue with merging this before `wrapOBS` and the plugin changes make it to stable?

### Checklist

- [X] Change is backwards compatible.

- [X] Code formatted with `./format`.

- [X] Code tested through `nix-shell --pure tests -A run.all`.

- [X] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [X] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

### CODE REVIEW

File: modules/programs/obs-studio.nix [hunk 1]
```diff
@@ -37,19 +22,27 @@ in {
...
   };
 
   config = mkIf cfg.enable {
-    home.packages = [ cfg.package ];
-
-    xdg.configFile."obs-studio/plugins" =
-      mkIf (cfg.plugins != [ ]) { source = mkPluginEnv cfg.plugins; };
+    home.packages = [ cfg.finalPackage ];
+    programs.obs-studio.finalPackage =
+      pkgs.wrapOBS.override { obs-studio = cfg.package; } {
```
Comment (deviant) [inline modules/programs/obs-studio.nix:44]: Quick question: if it's only used in this one place and can't be modified nor seen, why not just inline it into the `home.packages = [ ... ];`?

**Review by deviant — APPROVED** [review]

File: modules/programs/obs-studio.nix (hunk 1 above)
Comment (06kellyjac) [reply modules/programs/obs-studio.nix:44]: I'll leave that to someone more familiar with the `home-manager` repo

`finalPackage` isn't used anywhere else for the `neovim` module either

Comment (berbiche) [reply modules/programs/obs-studio.nix:44]: The `finalPackage` pattern is really nice to use when something is wrapped.
This pattern is used with Emacs, Neovim and others.

**Review by berbiche — APPROVED** [review]
Thanks for the contribution!

image
<https://nixpk.gs/pr-tracker.html?pr=125308>

Safe to merge!

File: modules/programs/obs-studio.nix (hunk 1 above)
Comment (deviant) [reply modules/programs/obs-studio.nix:44]: What's it useful for? The usual way to do this (if a non-anonymous binding is desired, which is generally only done if using the binding in multiple places) is with `let`:

```nix
let
  cfg = options.programs.obs-studio;
  finalPackage = wrapOBS.override { obs-studio = cfg.package; } {
    inherit (cfg) plugins;
  };
in {
  /* ... */
  config = mkIf cfg.enable {
    home.packages = [ finalPackage ];
  };
}
```

Being read-only, this isn't possible to set from anywhere else, even using `mkForce`:

```
nix-repl> with lib; (evalModules {
            modules = [
              {
                options.foo = mkOption {
                  type = types.int;
                  readOnly = true;
                };
                config.foo = 1;
              }
              {
                foo = lib.mkForce 2;
              }
            ];
          }).config.foo
error: The option `foo' is read-only, but it's set multiple times. Definition values:
- In `<unknown-file>': 2
- In `<unknown-file>': 1
```

If this is to make the package available from *outside* the module, it wouldn't be marked hidden: `visible = false` implies it's internal to the module/an implementation detail, and shouldn't be considered stable for external consumers. Are you sure this isn't just something that's been copy/pasted between modules a couple of times?

**Comment by JonathanLorimer** [comment]
Just wanted to say thank you for dealing with this so quickly! Everything seems to be working for me now.

**Comment by 06kellyjac** [comment]
Glad to help 🙂

**Pull request merged**